data/transactions/logic/resources.go

```diff
@@ -0,0 +1,359 @@
+	// tx0 mentions an account A, and tx1 mentions an ASA X, that does _not_
+	// make the holding AX available
+	sharedHoldings map[ledgercore.AccountAsset]struct{}
+	sharedLocals   map[ledgercore.AccountApp]struct{}
+
+	// boxes are all of the top-level box refs from the txgroup. Most are added
+	// during NewEvalParams(). refs using 0 on an appl create are resolved and
+	// added when the appl executes. The boolean value indicates the "dirtiness"
+	// of the box - has it been modified in this txngroup? If yes, the size of
+	// the box counts against the group writeBudget. So delete is NOT a dirtying
+	// operation.
+	boxes map[boxRef]bool
+
+	// dirtyBytes maintains a running count of the number of dirty bytes in `boxes`
+	dirtyBytes uint64
+}
+
+func (r *resources) shareHolding(addr basics.Address, id basics.AssetIndex) {
+	r.sharedHoldings[ledgercore.AccountAsset{Address: addr, Asset: id}] = struct{}{}
+}
+
+func (r *resources) shareAccountAndHolding(addr basics.Address, id basics.AssetIndex) {
+	r.sharedAccounts[addr] = struct{}{}
+	if id != 0 {
+		r.sharedHoldings[ledgercore.AccountAsset{Address: addr, Asset: id}] = struct{}{}
+	}
+}
+
+func (r *resources) shareLocal(addr basics.Address, id basics.AppIndex) {
+	r.sharedLocals[ledgercore.AccountApp{Address: addr, App: id}] = struct{}{}
+}
+
+// In the fill* and allows* routines, we pass the header and the fields in
+// separately, even though they are pointers into the same structure. That
+// prevents dumb attempts to use other fields from the transaction.
+
+func (r *resources) fill(tx *transactions.Transaction, ep *EvalParams) {
+	switch tx.Type {
+	case protocol.PaymentTx:
+		r.fillPayment(&tx.Header, &tx.PaymentTxnFields)
+	case protocol.KeyRegistrationTx:
+		r.fillKeyRegistration(&tx.Header)
+	case protocol.AssetConfigTx:
+		r.fillAssetConfig(&tx.Header, &tx.AssetConfigTxnFields)
+	case protocol.AssetTransferTx:
+		r.fillAssetTransfer(&tx.Header, &tx.AssetTransferTxnFields)
+	case protocol.AssetFreezeTx:
+		r.fillAssetFreeze(&tx.Header, &tx.AssetFreezeTxnFields)
+	case protocol.ApplicationCallTx:
+		r.fillApplicationCall(ep, &tx.Header, &tx.ApplicationCallTxnFields)
+	case protocol.StateProofTx:
+		// state proof txns add nothing to availability (they can't even appear
+		// in a group with an appl. but still.)
+	default:
+		panic(tx.Type)
+	}
+}
+
+func (cx *EvalContext) allows(tx *transactions.Transaction, calleeVer uint64) error {
+	switch tx.Type {
+	case protocol.PaymentTx, protocol.KeyRegistrationTx, protocol.AssetConfigTx:
+		// these transactions don't touch cross-product resources, so no error is possible
+		return nil
+	case protocol.AssetTransferTx:
+		return cx.allowsAssetTransfer(&tx.Header, &tx.AssetTransferTxnFields)
+	case protocol.AssetFreezeTx:
+		return cx.allowsAssetFreeze(&tx.Header, &tx.AssetFreezeTxnFields)
+	case protocol.ApplicationCallTx:
+		return cx.allowsApplicationCall(&tx.Header, &tx.ApplicationCallTxnFields, cx.version, calleeVer)
+	default:
```

Comment:
What about rekey

Reply:
rekey is not a transaction type, it's a field on all transactions.  the argument feels like an account, but it's really only a 32 byte value - that is, the existence of the value as an account doesn't matter, so it is not looked up. So it should not be added to the availability set.